casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -261,6 +261,17 @@ object BlockReceiver {
               hasAllDeps = pendingRequests.isEmpty &&
                 block.justifications.forall(dag.contains)
 
+              // If replay was interrupted, block was stored but not validated or added to the DAG.
+              // Validation needs to be restarted for such blocks
+              unvalidatedParents <- block.justifications.filterA { hash =>
+                                     BlockStore[F].contains(hash) &&^ dag.contains(hash).pure.not
+                                   }
+              _ <- unvalidatedParents.traverse { hash =>
+                    state.modify(_.beginStored(hash)) *>
+                      state.modify(_.endStored(hash, List.empty)) *>
```

Comment:
Between these calls of `state.modify` another thread can also change the state. Use only one `state.modify` and inside made all changes atomically to the state. Also if you don't care about result from state modification you can use `state.update`.